sbot/servo_board.py

```diff
@@ -0,0 +1,314 @@
+        response = self._serial.query('*IDN?')
+        return BoardIdentity(*response.split(':'))
+
+    @log_to_debug
+    def status(self) -> tuple[bool, bool]:
+        """
+        Get the board's status.
+
+        :return: A tuple of the watchdog fail and pgood status.
+        """
+        response = self._serial.query('*STATUS?')
+
+        data = response.split(':')
+        watchdog_fail = (data[0] == '1')
+        pgood = (data[1] == '1')
+
+        return watchdog_fail, pgood
+
+    @log_to_debug
+    def reset(self) -> None:
+        """
+        Reset the board.
+
+        This will disable all servos.
+        """
+        self._serial.write('*RESET')
+
+    @property
+    @log_to_debug
+    def current(self) -> float:
+        """
+        Get the current draw of the board.
+
+        This only includes the servos powered through the main port, not the aux port.
+
+        :return: The current draw of the board in amps.
+        """
+        response = self._serial.query('SERVO:I?')
+        return float(response) / 1000
+
+    @property
+    @log_to_debug
+    def voltage(self) -> float:
+        """
+        Get the voltage of the on-board regulator.
+
+        :return: The voltage of the on-board regulator in volts.
+        """
+        response = self._serial.query('SERVO:V?')
+        return float(response) / 1000
+
+    def _cleanup(self) -> None:
+        """
+        Reset the board and disable all servos on exit.
+
+        This is registered as an exit function.
+        """
+        try:
+            self.reset()
+        except Exception:
+            logger.warning(f"Failed to cleanup servo board {self._identity.asset_tag}.")
+
+    def __repr__(self) -> str:
+        return f"<{self.__class__.__qualname__}: {self._serial}>"
+
+
+class Servo:
+    """
+    A class representing a servo on the servo board.
+
+    :param serial: The serial wrapper to use to communicate with the board.
+    :param index: The index of the servo on the board.
+    """
+    __slots__ = ('_serial', '_index', '_duty_min', '_duty_max')
+
+    def __init__(self, serial: SerialWrapper, index: int):
+        self._serial = serial
+        self._index = index
+
+        self._duty_min = START_DUTY_MIN
+        self._duty_max = START_DUTY_MAX
+
+    @log_to_debug
+    def set_duty_limits(self, lower: int, upper: int) -> None:
+        """
+        Set the pulse on-time limits of the servo.
+
+        These limits are used to map the servo position to a pulse on-time.
+
+        :param lower: The lower limit of the servo pulse in µs.
+        :param upper: The upper limit of the servo pulse in µs.
+        :raises TypeError: If the limits are not ints.
+        :raises ValueError: If the limits are not in the range 500 to 4000.
+        """
+        if not (isinstance(lower, int) and isinstance(upper, int)):
+            raise TypeError(
+                f'Servo pulse limits are ints in µs, in the range {DUTY_MIN} to {DUTY_MAX}'
+            )
```

Comment:
Suggestion: Instead of a duplicate type check, perhaps we can keep the range check below, and implicitly cast them when storing them below?

Reply:
Still need the type checking so we get a nice error if you can't compare to an int.